_posts/2020-08-02-prediction-of-medical-costs/2020-08-02-prediction-of-medical-costs.Rmd

```diff
@@ -0,0 +1,600 @@
+
+
+```{r echo=FALSE, eval=longrun}
+scores = bmr$score(measures) %>%
+  select(Learner = learner_id, Iteration = iteration, MSE = regr.mse)
+
+png("bmr_plot.png")
+plot_bmr = autoplot(bmr) +
+  labs(x = "Learner", y = "MSE", title = "MSE-Score for each learner over the performance estimates") +
+  theme_bw()
+dev.off()
+```
+
+![Benchmarking Boxplot](bmr_plot.png)
+
+Benchmarking generates scores for each learner and each outer loop iteration.
+Drawing boxplots over the performance estimates - here the Mean Squared Error - shows that kNN seems to work fine but produces some variation in performance estimates.
+The Linear Model doesn't deliver satisfactory results and produces large MSEs compared to the other candidate learners.
+Random Forest appears to be the best model producing accurate predictions in terms of the MSE.
+Finally, Decision Trees produce performance estimates similar to those of kNN, with even smaller variability.  
+
+One of our initial questions was: Which learner qualifies best to predict insurance charges - that is, as accurately as possible?  
+
+To come to answer this question, we do not necessarily have to study the boxplots - though it might be helpful.
+Instead, we can access the aggregated performance estimates for each learner, directly.
+In the graphic below the learners are ranked according to our three performance measures:
+MSE, MAE and $R^2$.
+The inverse value of $R^2$ has been computed and plotted to allow for easy comparison; improvements are indicated by lower values in each of the three plots.
+
+```{r, eval=longrun, echo=FALSE}
+tab = bmr$aggregate(measures)
+
+ranks = tab[, .(learner_id, regr.mse, rank_mse = rank(regr.mse), regr.mae, rank_mae = rank(regr.mae), inv_regr.rsq = 1 / regr.rsq, rank_rsq = rank(-regr.rsq))]
+
+saveRDS(ranks, "ranks.rds")
+```
+
+```{r tabrds, echo=FALSE, eval=!longrun}
+ranks = readRDS("ranks.rds")
+```
+
+
+```{r echo=FALSE, eval=TRUE}
+ranks = ranks %>% pivot_longer(c(regr.mse, regr.mae, inv_regr.rsq), names_to = "key", values_to = "value")
+ranks$key = factor(ranks$key, levels = c("regr.mse", "regr.mae", "inv_regr.rsq"))
+
+# Visualization
+ggplot(data = ranks) +
+  geom_point(aes(x = key, y = value, col = as.factor(learner_id)), size = 5) +
+  facet_wrap(~key,
+    scales = "free",
+    labeller = as_labeller(c("regr.mse" = "MSE", "regr.mae" = "MAE", "inv_regr.rsq" = "Inverse of R^2"))) +
+  theme_bw() +
+  theme(axis.text.x = element_blank(), axis.title = element_blank(), legend.position = "right") +
+  scale_color_discrete(labels = c(
+    "regr.kknn.tuned" = "kNN (tuned)",
+    "regr.lm" = "LM",
+    "regr.ranger.tuned" = "Ranger (tuned)",
+    "regr.rpart.tuned" = "CART (tuned)")) +
+  labs(col = "Learner")
+```
+
+
+But what about the hyperparameters?
+We just learned that from nested resampling we obtain five different sets of optimal combinations:
+Which one of them shall we use for the final model?  
+
+The answer is: none of them.
```

Comment:
Explain why we do the outer loop stuff then. I.e. in order to obtain unbiased measure of generalization error.